### PR TITLE
Implemented connecting to the rooms.

### PR DESCRIPTION
**Pull request by Subv**
Nicknames and MAC addresses are not yet negotiated.
Only wifi beacon frames are retransmitted for the moment.

Moved the header files to a tunnler/ subdirectory.

There are still some concurrency issues that i'd like help with, namely in RoomMember.

### CODE REVIEW

File: include/tunnler/room_member.h
Comment (Subv) [inline include/tunnler/room_member.h]: Maybe wrap this in std::atomic?

File: include/tunnler/room.h
Comment (JayFoxRox) [inline include/tunnler/room.h]: nit: If you rename it to `GetState` also say "state" here

File: src/room_member.cpp
Comment (Subv) [inline src/room_member.cpp]: I think this is a race condition with Leave()

File: include/tunnler/room_member.h
Comment (JayFoxRox) [inline include/tunnler/room_member.h]: Why isn't this const anymore?

Comment (Subv) [reply include/tunnler/room_member.h]: Because there's a very very awful
````
stream.Write(reinterpret_cast<char*>(wifi_packet.data.data()), wifi_packet.data.size());
````

which would have to become either
````
stream.Write(const_cast<char*>(reinterpret_cast<const char*>(wifi_packet.data.data())), wifi_packet.data.size());
````
or
````
stream.Write((char*)wifi_packet.data.data(), wifi_packet.data.size());
````

Suggestions?

File: include/tunnler/room_member.h
Comment (JayFoxRox) [inline include/tunnler/room_member.h]: spelling: "received" - Maybe also put the word "Data" in backticks / quotes?
Same with Beacon line below

File: include/tunnler/room.h
Comment (B3n30) [inline include/tunnler/room.h]: Doesn't return anything

File: src/room.cpp
Comment (JayFoxRox) [inline src/room.cpp]: In the long run this should be moved to the server creation / memberlimit. Temporarily it's fine

File: src/room.cpp
Comment (JayFoxRox) [inline src/room.cpp]: explicit braces / scope plox.

File: src/room.cpp
Comment (JayFoxRox) [inline src/room.cpp]: We should assert that we closed successfully. I'm not sure if raknet destroying the instance returns anything / could fail. But it would be quite horrendous if a host closes the server, returns a "Closed" state but might still have some socket open.

If raknet is fail-safe in this regard this is fine.

File: src/room.cpp
Comment (JayFoxRox) [inline src/room.cpp]: explicit scope once more

File: src/room.cpp
Comment (Subv) [reply src/room.cpp]: I have no idea if that's the case, what would we even check for here?

File: src/room.cpp
Comment (JayFoxRox) [inline src/room.cpp]: This should have a fallback ASAP so we see any raknet packet we might have to worry about in the future

File: src/room.cpp
Comment (JayFoxRox) [inline src/room.cpp]: nit: didn't we decide to make these `RELIABLE` but unordered?

File: src/room_member.cpp
Comment (JayFoxRox) [inline src/room_member.cpp]: more scope fun

File: src/room.cpp
Comment (Subv) [reply src/room.cpp]: true, leftover from testing

File: src/room_member.cpp
Comment (JayFoxRox) [inline src/room_member.cpp]: *cough* - this time it actually even confused me because the indentation is so unexpected

File: src/room_member.cpp
Comment (JayFoxRox) [inline src/room_member.cpp]: Sounds like we should just create our own class / struct for this?

File: src/room.cpp
Comment (Subv) [reply src/room.cpp]: what would the fallback do? we'll be getting *tons* of raknet messages that we don't care about (incoming_connection notification, etc)

File: src/room_member.cpp
Comment (JayFoxRox) [inline src/room_member.cpp]: same thing again: not sure if RELIABLE only?

File: src/room_member.cpp
Comment (Subv) [reply src/room_member.cpp]: that's imo too much work for little payoff, we can just pop_front() the data queue like we do for the beacon queue

**Comment by Subv** [comment]
Addressed some @JayFoxRox comments.

File: src/room_member.cpp
Comment (JayFoxRox) [inline src/room_member.cpp]: Aren't we still Joining?
We are now halfway through the handshake.
I'd expect us to be completly joined once we have received confirmation for our nickname + mac address

File: src/room_member.cpp
Comment (JayFoxRox) [inline src/room_member.cpp]: Did you check which events are automaticly closing the connection and which don't?
If the server is full I'd expect raknet to internally have closed the connection. if it doesn't it might not do so here either? (maybe it will just try to reconnect when continuing sending data?)

File: src/room_member.cpp
Comment (Subv) [reply src/room_member.cpp]: See the TODO. This was basically testing code.

File: include/tunnler/room_member.h
Comment (JayFoxRox) [reply include/tunnler/room_member.h]: I'd do the casting back and forth or even create a copy of the packet, just to not give in to RakNet shenannigans.

Before sending that E-Mail I might check if there is a PR / reason for this and tell them to fix it too :P

File: src/room_member.cpp
Comment (JayFoxRox) [reply src/room_member.cpp]: The TODO comment only says to send the packet. What I'm saying is: wait for a response.
So the state should not be modified in this function at all.

File: src/room_member.cpp
Comment (JayFoxRox) [inline src/room_member.cpp]: spelling

File: src/room_member.cpp
Comment (JayFoxRox) [inline src/room_member.cpp]: Spelling.. still [github is being stupid]

File: src/room_member.cpp
Comment (JayFoxRox) [inline src/room_member.cpp]: I believe offline raknet peers do some things such as responding to pings.
If that is the case, we should consider only opening the peers once actual connection stuff is about to happen maybe?

File: src/room_member.cpp
Comment (Subv) [inline src/room_member.cpp]: This Shutdown() call will cause the peer->Receive() loop to stop immediately, is that desired? Should we care about queued packets?

Comment (Subv) [reply src/room_member.cpp]: A possible fix would be to have a lock here and inside the while (IsConnected()) loop. That ensures that we have processed all queued packets before shutting down, similar to how it's done in the Room class.

File: src/room.cpp
Comment (B3n30) [inline src/room.cpp]: Shouldn't we send a ID_DISCONNECTION_NOTIFICATION to the clients?

File: src/room.cpp
Comment (B3n30) [inline src/room.cpp]: Shouldn't this be before the server is closed. I know for now it won't make a difference, but maybe for later implementations. It's just logically to end the ServerLoop before destroying the server.

File: src/room.cpp
Comment (B3n30) [inline src/room.cpp]: In the Raknet examples a Sleep(30 / u_sleep(30*1000) is used to "keep Raknet responsive". I don't know if this is necessary.

File: src/room_member.cpp
Comment (B3n30) [inline src/room_member.cpp:112]: In the Raknet examples a Sleep(30 / u_sleep(30*1000) is used to "keep Raknet responsive". I don't know if this is necessary.

File: include/tunnler/assert.h
Comment (B3n30) [inline include/tunnler/assert.h:1]: Should we add tunnler. At least its 2017... :D

File: include/tunnler/room.h
Comment (B3n30) [inline include/tunnler/room.h:35]: We now have Room::MemberList and RoomMember::MemberList, both holding almost the same information. I know technically there is no problem with that, but going to the code sometimes it's hard to distinguish which one is meant. Maybe rename one to ServerMemberList or similar or use a std::map<RakNet::AddressOrGUID, RoomMember::MemberInformation> since we have to store the complete MemberInformation on the Server anyway.

File: include/tunnler/room_member.h
Comment (B3n30) [inline include/tunnler/room_member.h:126]: member_information_list ?

File: include/tunnler/tunnler.h
Comment (B3n30) [inline include/tunnler/tunnler.h]: Maybe just use the broadcast mac for that, since this should be the only mac address a client shouldn't have. 0:0:0:0:0:0 is as far as i know a legit mac address.

File: include/tunnler/room_member.h
Comment (B3n30) [inline include/tunnler/room_member.h:131]: Maybe just store MemberInformation for this client, thus MAC adress is stored there too.

File: src/room.cpp
Comment (B3n30) [inline src/room.cpp]: unnecessary space

File: src/room.cpp
Comment (B3n30) [inline src/room.cpp:125]: If my suggestion about NoPreferredMac=FF:FF:FF:FF:FF:FF isn't considered here should't be a check that address not equals the broadcast address. I can't think of any other special macs but broadcast shoudn't be vaild

File: src/room_member.cpp
Comment (B3n30) [inline src/room_member.cpp:23]: Why not just call Leave? Since this destructor my be called even if I'm connected (for example just closing citra while connected)

File: include/tunnler/room.h
Comment (B3n30) [inline include/tunnler/room.h:38]: At least Destroy() should be called here

**Review by B3n30 — CHANGES_REQUESTED** [review]
I might not got everything.
Some are just questions or notes to think about

File: include/tunnler/assert.h
Comment (JayFoxRox) [reply include/tunnler/assert.h:1]: imo: Nope, it's not our file

File: src/room_member.cpp
Comment (JayFoxRox) [reply src/room_member.cpp:23]: imo it should be part of the Application / Citra cleanup to worry about these things.
The current behaviour is fine

File: src/room.cpp
Comment (JayFoxRox) [reply src/room.cpp]: isn't that part of Shutdown? *assumes*

File: include/tunnler/room.h
Comment (JayFoxRox) [reply include/tunnler/room.h:38]: same as for RoomMember: Leave this kind of behaviour to applications

File: include/tunnler/room.h
Comment (JayFoxRox) [reply include/tunnler/room.h:35]: I didn't read the code yet [just your review comments], but introducing "Server" would be even more confusing.
The general concept of naming these functions was that a non-network person or even layman would get these terms - this will also be reflected by the GUI.

File: src/room.cpp
Comment (JayFoxRox) [reply src/room.cpp]: lol LC?!

File: src/room.cpp
Comment (B3n30) [reply src/room.cpp]: You are right. Just checked the doc

File: src/room.cpp
Comment (B3n30) [reply src/room.cpp]: sorry. I meant unnecessary line :-P

File: include/tunnler/room_member.h
Comment (JayFoxRox) [reply include/tunnler/room_member.h:126]: I think the current name is fine: "information" is non-countable. while this doesn't imply a plural from the naming, it does from the context (and so does the comment [which should probably be renamed to say ~"all room-members in the joined room" instead of client])

File: include/tunnler/room_member.h
Comment (JayFoxRox) [reply include/tunnler/room_member.h:131]: Not sure. I feel that the MemberInformation / MemberList stuff is a user-facing (API) based structure.

Consider network activity: If we count network activity (sent WifiPackets / s per RoomMember) to show in the frontend* we'd suddenly have useless fields stored in the RoomMember.

Also it adds a layer of indirection which might make the code slightly more complicated for no reason.

*(which I'm intending so network spammers could be found more easily)

File: include/tunnler/room.h
Comment (B3n30) [reply include/tunnler/room.h:38]: Imo this couldn't be handled by the applications, since the destructor is called "automatically" in some cases. Thus closing the room/room_member and sending the disconnect message should be handled...

File: src/room_member.cpp
Comment (JayFoxRox) [inline src/room_member.cpp:69]: Given B3N30s latest comment: It should only be sensical that filtering for `mac_address == NoPreferredMac / BroadcastMac` should disable the filter action (for mac only). Keep this in mind for the logic and documentation.

(Add a TODO?)

File: src/room.cpp
Comment (JayFoxRox) [reply src/room.cpp]: I get that - but this is something lioncash (= LC) would say during review (and LC reviews *everything*).
He's very "uptight" (for a lack of a better term) about keeping code short by removing useless empty lines :P

File: src/room.cpp
Comment (JayFoxRox) [reply src/room.cpp]: the room_thread is a `unique_ptr` / doesn't the thread get destroyed automaticly?
This is only do to things gracefully - we already assert in the other case so API users should know that shit is about to go down (we should just document the fact that rooms must be `destroy()`d before destroying the object).
However, maybe we *must* forcefully [not the same as `leave()`!] kill the thread to avoid race conditions where the room_thread will use an already closed / destroyed RakPeer - but we might be doing that already, I didn't check

**Comment by JayFoxRox** [comment]
We are discussing minuscule things by now. I feel like this has been discussed enough. 
So I suggest keeping "nits" to yourself at this point and only address actual design issues.

@Subv please handle the latest review comments; then make sure this code is compilable so we can merge this to master asap (just handle the remaining review comments). We should have a stable master as base to work on for more important things.

(If remaining design or code issues are found we can discuss them in GitHub issues or add comments after the merge)

File: include/tunnler/room_member.h
Comment (JayFoxRox) [reply include/tunnler/room_member.h]: To resolve this before the merge: please make it accept a const ref. I feel like this is part of a stable API so it would suck to change it later

File: src/room.cpp
Comment (B3n30) [reply src/room.cpp]: Since Subv reports it works without the sleep I will skip it for now

File: src/room_member.cpp
Comment (B3n30) [reply src/room_member.cpp:112]: Will skip this, too

File: src/room.cpp
Comment (JayFoxRox) [reply src/room.cpp]: It might work for his platform.. - This is practically a busyloop.
We should fix this. At least we should have a reminder about it in an issue or something.

File: CMakeLists.txt
Comment (JayFoxRox) [inline CMakeLists.txt]: Any reason we don't use the cmake way of doing this?
(or use whatever Citra uses :P )

Comment (B3n30) [reply CMakeLists.txt]: Only my stupidity and my inability of writing cmake. I will just look what citra uses and add it here

File: src/room.cpp
Comment (JayFoxRox) [inline src/room.cpp]: Not sure about this change, I guess it's fine, but I feel like we should continue handling packets even during shutdown.
Thread closure (setting state + joining thread) should probably happen after shutdown, before destroying the instance.

//edit: I'll add an issue once we merged this [or do it now, it fits with the next PR]

File: src/room.cpp
Comment (JayFoxRox) [reply src/room.cpp:125]: resolved

File: src/room_member.cpp
Comment (JayFoxRox) [reply src/room_member.cpp:23]: resolved

**Comment by JayFoxRox** [comment]
This looks good now, basicly this C14 thing and an edge-case. Merge please :)
We can then do follow up PRs

File: CMakeLists.txt
Comment (B3n30) [reply CMakeLists.txt]: addressed

File: src/room.cpp
Comment (B3n30) [reply src/room.cpp]: I think since we close the room, receiving packets isn't needed any more, and the way I arranged it, allows the room/server to work without a mutex

**Comment by B3n30** [comment]
For me LGTM...

File: src/room.cpp
Comment (JayFoxRox) [reply src/room.cpp]: In case of RoomMember we'd still get confirmation that we successfully disconnected etc. (I believe)
I guess the same will hold true for Room.

Comment (Subv) [reply src/room.cpp]: This order of things is actually ok, the thread will keep receiving packets until there's nothing more to receive, then exit the inner loop and then exit the outer loop (because state = Closed), after that we shut down the instance (which should tell the clients that we're shutting down, if i'm reading the docs correctly)

**Review by B3n30 — APPROVED** [review]

File: src/room_member.cpp
Comment (Subv) [reply src/room_member.cpp:69]: The to-do is already there, see L73

**Comment by Subv** [comment]
LGTM, let's merge it and iterate on it with further PRs

**Pull request merged**
